Provider does not work with physical PIV tokens

**Describe the bug**
Fails to access keys on the token

**To Reproduce**
Do
```
openssl cms -aes256 -decrypt -binary -inform PEM -in /tmp/derive.27490.text.cms -out /tmp/derive.27490.text.dec -inkey "pkcs11:id=%00%9d;type=private"
```
and observe response
```
Could not open file or uri for loadingCould not read key etc. of signing key from pkcs11:id=%00%9d;type=private
40E6BC57F87F0000:error:80000002:system library:file_open:No such file or directory:providers/implementations/storemgmt/file_store.c:265:calling stat(pkcs11:id=%00%9d;type=private)
40E6BC57F87F0000:error:1608010C:STORE routines:inner_loader_fetch:unsupported:crypto/store/store_meth.c:353:No store loader found. For standard store loaders you need at least one of the default or base providers available. Did you forget to load them? Info: Global default library context, Scheme (pkcs11 : 0), Properties (<null>)
```

**Expected behavior**
Provider accessed the key on the token, and correctly decrypted the expected value.

**Operating environment (please complete the following information):**
 - OS: MacOS Ventura
 - Version 13.2
 - OpenSSL ~~3.0.7~~ 3.0.8 and 3.2.0-dev

**Token and application used (please complete the following information):**
 - Device: Yubikey-4, Yubikey-5, DoD CAC
 - PKCS11 Driver version: OpenSC master
 - Application: OpenSSL
 - Versions tested: 3.0.7, 3.0.8, and 3.2.0-dev

**Additional context**
Examples in `tests/` are insufficient - trying key URI snatched from `testvars` `BASEURI` and such did not help.
**Update**: following `HOWTO.md` and ensuring all the config files are in sync, URI `pkcs11:id=%0x;type=private` (where `x` is {1,2,3} for `PIV Auth` (slot 9a), `SIGN` (slot 9c), and `KEY MAN` (slot 9d) private key correspondingly) points at the correct key.

Some more datapoints:
```
$ openssl dgst -engine pkcs11 -keyform engine -sign "pkcs11:manufacturer=piv_II;id=%02;type=private" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -out ~/src/t.c.sig ~/src/t.c
Engine "pkcs11" set.
Enter PKCS#11 token PIN for XXXXXXXXXX:
$ ll ~/src/t.c
-rw-r--r--  1 ur20980  staff  1664 Feb  6  2012 /Users/ur20980/src/t.c
$ ll ~/src/t.c.sig
-rw-r--r--  1 ur20980  staff  256 Feb  8 23:29 /Users/ur20980/src/t.c.sig
$ openssl dgst -engine pkcs11 -keyform engine -verify "pkcs11:manufacturer=piv_II;id=%02;type=public" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -signature ~/src/t.c.sig ~/src/t.c
Engine "pkcs11" set.
Verified OK
$ 
```
Note: prompting for PIN **only** when private key operation is requested. Successful signing, successful verification of the signature. Demonstrated correct OpenSSL => engine => OpenSC => physical token (Yubikey-4).


With pkcs11-provider:
```
$ openssl dgst -sign "pkcs11:manufacturer=piv_II;id=%02;type=private" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -out ~/src/t.c.prov_sig ~/src/t.c
Enter pass phrase for PKCS#11 Token (Slot 4 - Yubico YubiKey OTP+FIDO+CCID):
Read error in /Users/ur20980/src/t.c
40F6064DF87F0000:error:42000001:pkcs11:p11prov_sig_operate_init:reason(1):signature.c:716:Failed to open session on slot 4
$ ll ~/src/t.c.prov_sig
-rw-r--r--  1 ur20980  staff  0 Feb  8 23:32 /Users/ur20980/src/t.c.prov_sig
$ openssl dgst -verify "pkcs11:manufacturer=piv_II;id=%02;type=public" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -signature ~/src/t.c.sig ~/src/t.c
Enter pass phrase for PKCS#11 Token (Slot 4 - Yubico YubiKey OTP+FIDO+CCID):
Verification failure
40F6064DF87F0000:error:42000054:pkcs11:p11prov_GetOperationState:reason(84):./interface.gen.c:225:Error returned by C_GetOperationState
40F6064DF87F0000:error:420000C0:pkcs11:p11prov_VerifyFinal:reason(192):./interface.gen.c:644:Error returned by C_VerifyFinal
$ 
```
Provider:
- failed to sign;
- failed to verify somebody else's good signature;
- prompted for a pin when it shouldn't.

Here's OpenSC verifying the signature made by OpenSSL (using engine) above:
```
$ pkcs11-tool --module /Library/OpenSC/lib/opensc-pkcs11.so --verify -m SHA384-RSA-PKCS-PSS -y pubkey -d 02 --signature-file ~/src/t.c.sig -i ~/src/t.c
Using slot 1 with a present token (0x4)
Using signature algorithm SHA384-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA384, mgf=MGF1-SHA384, salt_len=48 B
Signature is valid
$ 
```
Observe that OpenSC (and OpenSSL) does not bother user to enter a pin for publicly-available pubkey on the token. Can you please make the provider behave the same way?

> @mouse07410 this is really strange, it is the same error as before, but it shouldn't happen with the patch now applied

**I take it back! My fault!** I ~~stupidly~~ forgot a small step - _actually **installing_** the new provider.

With the new provider (#189 applied):
```
$ openssl pkeyutl -sign -inkey "pkcs11:manufacturer=piv_II;id=%02;type=private" -pkeyopt rsa_padding_mode:pkcs1 -in data_to_sign2.bin -out data_to_sign2.bin.sig2
Enter pass phrase for PKCS#11 Token (Slot 4 - Yubico YubiKey OTP+FIDO+CCID):
$ 
$ openssl pkeyutl -verify -pubin -inkey "pkcs11:manufacturer=piv_II;id=%02;type=public" -pkeyopt rsa_padding_mode:pkcs1 -in data_to_sign2.bin -sigfile data_to_sign2.bin.sig2
Enter pass phrase for PKCS#11 Token (Slot 4 - Yubico YubiKey OTP+FIDO+CCID):
Signature Verified Successfully
$ 
```
That's a good progress!

However, there still are problems:
```
$ openssl dgst -sign "pkcs11:model=PKCS%2315%20emulated;id=%02;type=private" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -out ~/src/t.c.prov-sig ~/src/t.c
Enter pass phrase for PKCS#11 Token (Slot 4 - Yubico YubiKey OTP+FIDO+CCID):
Error signing data
40F6064DF87F0000:error:42000054:pkcs11:p11prov_GetOperationState:reason(84):./interface.gen.c:225:Error returned by C_GetOperationState
40F6064DF87F0000:error:42000007:pkcs11:p11prov_SignFinal:Invalid or improper arguments were provided to the invoked function:./interface.gen.c:579:Error returned by C_SignFinal
```

More details later, probably on Sunday. In the meanwhile. @dengert probably could test with his tokens (and my script ;), and @simo5 might be able to test with his Yubi 4Nano, if it has PIV applet provisioned.

--- a/src/kdf.c
+++ b/src/kdf.c
@@ -165,7 +165,7 @@ static int p11prov_hkdf_derive(void *ctx, unsigned char *key, size_t keylen,
     if (key_size != keylen) {
         ret = CKR_GENERAL_ERROR;
         P11PROV_raise(hkdfctx->provctx, ret,
-                      "Expected derived key of len %lz, but got %lu", keylen,
+                      "Expected derived key of len %zu, but got %lu", keylen,
                       key_size);
         return RET_OSSL_ERR;
     }

--- a/src/session.c
+++ b/src/session.c
@@ -959,9 +959,8 @@ static CK_RV slot_login(P11PROV_SLOT *slot, P11PROV_URI *uri,
     /* we have a locked session, check that it is ok */
     ret = session_check(session, flags, &state);
     if (ret == CKR_OK) {
-        if (state == CKS_RW_PUBLIC_SESSION || state == CKS_RW_USER_FUNCTIONS
-            || state == CKS_RW_SO_FUNCTIONS) {
-            /* we have a valid logged in session */
+        if (state != CKS_RO_PUBLIC_SESSION) {
+            /* we seem to have a valid logged in session */
             goto done;
         }
     } else {

